File: onboarder/src/main/java/com/chyrta/onboarder/OnboarderPage.java
package com.chyrta.onboarder;

import android.graphics.drawable.Drawable;

import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;

public class OnboarderPage {

    public String title;
    public String description;
    public Drawable imageResource;
    @StringRes
    public int titleResourceId;
    @StringRes public int descriptionResourceId;
    @DrawableRes
    public int imageResourceId;
    @ColorRes
    public int titleColor;
    @ColorRes public int descriptionColor;
    @ColorRes public int backgroundColor;
    public float titleTextSize;
    public float descriptionTextSize;
    public boolean multilineDescriptionCentered;

    public OnboarderPage(String title, String description) {
        this.title = title;
        this.description = description;
        this.backgroundColor = R.color.black_transparent;
    }

    public OnboarderPage(String title, String description, int imageResourceId) {
        this.title = title;
        this.description = description;
        this.imageResourceId = imageResourceId;
        this.backgroundColor = R.color.black_transparent;
    }

    public OnboarderPage(String title, String description, Drawable imageResource) {
        this.title = title;
        this.description = description;
        this.imageResource = imageResource;
        this.backgroundColor = R.color.black_transparent;
    }

    public OnboarderPage(int title, int description) {
        this.titleResourceId = title;
        this.descriptionResourceId = description;
        this.backgroundColor = R.color.black_transparent;
    }

    public OnboarderPage(int title, int description, int imageResourceId) {
        this.titleResourceId = title;
        this.descriptionResourceId = description;
        this.imageResourceId = imageResourceId;
        this.backgroundColor = R.color.black_transparent;
    }

    public OnboarderPage(int title, int description, Drawable imageResource) {
        this.titleResourceId = title;
        this.descriptionResourceId = description;
        this.imageResource = imageResource;
        this.backgroundColor = R.color.black_transparent;
    }
}

File: onboarder/src/main/kotlin/com/chyrta/onboarder/OnboarderPage.kt
package com.chyrta.onboarder

import android.content.Context
import android.os.Parcelable
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.annotation.StyleableRes
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import kotlinx.parcelize.Parcelize

@Parcelize
data class OnboarderPage(
    /**
     * title
     */
    @StringRes val titleTextId: Int? = null,
    val titleText: CharSequence? = null,
    @ColorRes val titleColor: Int = R.color.white,
    val titleSize: Float? = null,
    @StyleableRes val titleFont: Int? = null,

    /**
     * description
     */
    @StringRes val descriptionTextId: Int? = null,
    val descriptionText: CharSequence? = null,
    @ColorRes val descriptionColor: Int = R.color.white,
    val descriptionSize: Float? = null,
    @StyleableRes val descriptionFont: Int? = null,

    /**
     * emblem
     */
    @DrawableRes
    val emblemDrawable: Int? = null,
    val emblemText: CharSequence? = null,
    val emblemTextSize: Float? = null,

    /**
     * other
     */
    val getActionLayoutFragment: (() -> Fragment)? = null,
    @ColorRes
    val backgroundColor: Int = R.color.black_transparent
) : Parcelable {

    companion object {
        const val EXTRA = "com.chyrta.onboarder.extra.OnboarderPage"
    }
}

fun Iterable<OnboarderPage>.backgroundColors(context: Context): List<Int> =
    map { ContextCompat.getColor(context, it.backgroundColor) }